benchmarks/profiling/readme.md

```diff
@@ -0,0 +1,177 @@
+```
+
+`perf_event_paranoid` setting is 1:
+- `-1`: Allow use of (almost) all events by all users
+      Ignore mlock limit after perf_event_mlock_kb without CAP_IPC_LOCK
+- `>= 0`: Disallow raw and ftrace function tracepoint - access
+- `>= 1`: Disallow CPU event access
+- `>= 2`: Disallow kernel profiling
+
+To make the adjusted `perf_event_paranoid` setting permanent preserve it
+in `/etc/sysctl.conf` (e.g. `kernel.perf_event_paranoid = <setting>`)
+
+### Run
+
+```
+./profile_deku.sh run
+```
+
+Profile deku node with perf:
+- Tear down the deku node
+- Setup deku environments
+- Start deku clusters and use `perf` to get the profiling data
+
+  The samples are saved in a `perf.data` file. This does involve CPU, file system, and disk overheads to save the samples to the file system for later processing.
+ - Sleep for 10 seconds
+
+<!------------------------------------------------------------------------>
+
+##  Read `perf.data`
+### [FlameGraph](https://github.com/brendangregg/FlameGraph)
+
+Read `perf.data` using FlameGraph. FlameGraph needs to be downloaded:
+ `git clone https://github.com/brendangregg/FlameGraph`
+
+Run this script only when `perf record` is **properly terminated** (a.k.a: `./profile_deku.sh run`).
+
+```
+./profile_deku.sh flame-graph
+```
+
+Wait for awhile to let the graph generated. The result in picture as `perf-deku-flamegraph.svg`. You can use internet browser to see it `firefox perf-deku-flamegraph.svg` or chrome, etc. 
+
+This is the generated flame-graph of deku-node running with 1012 blocks:
+[flame-graph.svg](https://github.com/lykimq/deku_docs/blob/main/benchmarking/graphs/perf-deku-flamegraph.svg)
+
+
+#### Understand Flamegraph
+It shows: 
+- x-axis: show the stack profile (the sample) population, sorted alphabetically
+- y-axis: stack depth, counting from zero at the bottom.
+- Each function (stack frame) is drawn as a rectangle, with the width relative to the number of samples. The wider a frame is, the more often it was present in the stacks.
+- The top edge shows what is on-CPU.
+- The beneath it is its ancestry.
+- The colors are usually not significant, picked randomly to differentiate frames.
+
+### Callgraph [gprof2dot](https://github.com/jrfonseca/gprof2dot)
+
+Read `perf.data` generated by Linux perf using `gprof2dot`.
+
+#### Setup
+
+Install on Debian/Ubuntu run first:
+
+```
+ sudo apt-get install python3 graphviz
+ sudo apt-get install python3-pip
+ sudo pip install gprof2dot
+```
+
+Wait for awhile to let the graph generated. The result in picture as `perf-deku-callgraph.png`. You can use internet browser to see it `firefox perf-deku-call-graph.png` or chrome, etc.
+
+This is the generated call-graph of deku-node running with 1012 blocks: 
+[call-graph.svg](https://github.com/lykimq/deku_docs/blob/main/benchmarking/graphs/perf-deku-callgraph.svg)
+
+Note that, it may happen that you cannot see this graph on a browser of Github, you can download this file and open it on local.
```

Comment:
You can remove this (you already said the first two paragraphs, the last line is not necessary IMO).